dcp_prototype/backend/wrangling/migrations/scripts/squish_files.py

```diff
@@ -0,0 +1,193 @@
+    number of files of each type that has been transferred over in the past to avoid
+    overwrites.
+
+    :param source_directory: The directory from which to copy files.
+    :param target_directory: The directory to which to copy and de-dupe files.
+    :param count_file: A file containing a dictionary that has the count of each file
+    processed. This count is used to append a suffix to each file name as it is copied
+    over.
+    :param checksums: A list of checksums of files that have already been processed.
+    :param clear_if_exists: If true and the target directory exists, then delete and
+    recreate the target directory while removing any existing files.
+    """
+
+    global COUNT_BY_FILE_TYPE
+
+    with open(count_file) as json_file:
+        COUNT_BY_FILE_TYPE = json.load(json_file)
+
+    if not os.path.isdir(target_directory):
+        os.mkdir(target_directory)
+    elif os.path.isdir(target_directory) and clear_if_exists:
+        rmtree(target_directory)
+    elif os.path.isdir(target_directory) and not clear_if_exists:
+        for file in os.listdir(target_directory):
+            with open(os.path.join(target_directory, file), "rb") as file_reader:
+                contents = file_reader.read()
+            checksums.append(hashlib.md5(contents).hexdigest())
+
+    # Prepopulate checksums with existing file checksums in target directory.
+    for filename in os.listdir(target_directory):
+        target_file_path = os.path.join(target_directory, filename)
+        with open(target_file_path, "rb") as file_reader:
+            contents = file_reader.read()
+        checksums.append(hashlib.md5(contents).hexdigest())
+
+    print(f"Processing directory {source_directory}")
+    _copy_files(source_directory, target_directory, COUNT_BY_FILE_TYPE, checksums)
+
+    # Store updated count files
+    if count_file:
+        json_file_pointer = open(count_file, "w")
+        json_file_pointer.write(json.dumps(COUNT_BY_FILE_TYPE))
+        json_file_pointer.close()
+
+
+def _copy_files(source_directory, target_directory, count_by_file_type, checksums):
+    """
+    Recursively copy over files as long as its checksum does not exist in the given list
+    of checksums and append the file with the count of the file type as designated by
+    the counts in `count_by_file_typ` if the file type exists.
+    """
+
+    for filename in os.listdir(source_directory):
+        # Skip hidden files
+        if filename.startswith("."):
+            continue
+
+        source_file_path = os.path.join(source_directory, filename)
+        if os.path.isdir(source_file_path):
+            print(f"Processing directory {source_file_path}")
+            _copy_files(
+                source_file_path, target_directory, count_by_file_type, checksums
+            )
+        elif os.path.isfile(source_file_path):
+            with open(source_file_path, "rb") as file_reader:
+                contents = file_reader.read()
```

Comment:
This will kind of blow up for bams and fastqs won't it?

Reply:
Yea we need to stream here. I'll work with Arathi on taking this files script work to completion. She has plenty on her plate with the metadata

Reply:
I'm curious. I've been running this just fine both on my laptop as well as a small-ish EC2 instance that Marcus set up. What is the blow up expected here?

Reply:
You're only working off of the pancreas dataset, which I'm assuming doesn't have that large of a fastq file. Other projects have fastq files that range from 20-54gb in size. You will run out of memory on most machines if you try to load this rather than streaming.

If you remember, this is the exact issue you encountered in the checksumming feature added to the hca cli 🙂

Reply:
The dcplib extract method lets you specify content type, I'm currently just loading in the metadata. I think for the data files it makes sense to copy directly between the s3 buckets rather than streaming to an ec2 instance/laptop and then uploading to s3.